src/pani/core.cljs

```diff
@@ -0,0 +1,161 @@
+(ns pani.core
+  (:refer-clojure :exclude [get-in set! reset! conj! swap! deref parents key])
+  (:require cljsjs.firebase
+            [clojure.string :as str]
+            [pani.registry :refer [register-listener]]))
+
+;; constants
+
+(def child-events
+  [:child-added
+   :child-changed
+   :child-moved
+   :child-removed])
+
+(def all-events
+  (conj child-events :value))
+
+;; utils
+
+;; FIXME: camel-case keys?
+;;        hydrate to custom vectors to preserve rich keys?
+;;        preserve sets (don't coerce to vector)
+;;        similarly preserve keywords as values
+
+(defn- hydrate [js-val]
+  (js->clj js-val :keywordize-keys true))
+
+(defn- serialize [clj-val]
+  (clj->js clj-val))
+
+(defn key
+  "Last segment in reference or snapshot path"
+  [ref]
+  (.key ref))
+
+(defn- value [snapshot]
+  (hydrate (.val snapshot)))
+
+(defn- kebab->underscore [keyword]
+  (-> keyword name (str/replace "-" "_")))
+
+(defn- underscore->kebab [string]
+  (-> string (str/replace "_" "-") keyword))
+
+(defn- korks->path [korks]
+  (if (sequential? korks)
+    (str/join "/" (map name korks))
+    (when korks (name korks))))
+
+(defn- wrap-snapshot [snapshot]
+  ;; TODO: enhance with snapshot protocol
+  [(key snapshot) (value snapshot)])
+
+;; references
+
+(defn connect
+  "Create a reference for firebase"
+  [url]
+  (js/Firebase. url))
+
+(defn get-in
+  "Obtain child reference from base by following korks"
+  [ref korks]
+  (let [path (korks->path korks)]
+    (if-not (seq path) ref (.child ref path))))
+
+;; FIXME: test to confirm this nil? is really necessary
+
+(defn parent
+  "Immediate ancestor of reference, if any"
+  [ref]
+  (let [p (and ref (.parent ref))]
+    (if (nil? (js->clj p)) nil p)))
+
+(defn parents
+  "Probably don't need this. Or maybe we want more zipper nav (siblings, in-order, etc)"
+  [ref]
+  (take-while identity (iterate parent (parent ref))))
+
+
+;; --------------------
+;; getters 'n setters
+
+(def undefined) ;; firebase methods do not take kindly to nil callbacks
```

Comment:
Certainly has an appeal, but also makes it sound callable (whereas it'll explode). We could actually just use the real `no-op`, the overhead would be pretty minor.

Another alternative, replace the `(or .. undefined)` forms with a `(presence ..)` (or better named) function, and keep undefined as hidden implementation detail.